substrate/primitives/runtime/src/traits/mod.rs

```diff
@@ -1474,8 +1551,67 @@ impl Dispatchable for () {
 	}
 }
 
+/// Dispatchable impl containing an arbitrary value which panics if it actually is dispatched.
+#[derive(Clone, Eq, PartialEq, Encode, Decode, RuntimeDebug, TypeInfo)]
+pub struct FakeDispatchable<Inner>(pub Inner);
+impl<Inner> From<Inner> for FakeDispatchable<Inner> {
+	fn from(inner: Inner) -> Self {
+		Self(inner)
+	}
+}
+impl<Inner> FakeDispatchable<Inner> {
+	/// Take `self` and return the underlying inner value.
+	pub fn deconstruct(self) -> Inner {
+		self.0
+	}
+}
+impl<Inner> AsRef<Inner> for FakeDispatchable<Inner> {
+	fn as_ref(&self) -> &Inner {
+		&self.0
+	}
+}
+
+impl<Inner> Dispatchable for FakeDispatchable<Inner> {
+	type RuntimeOrigin = ();
+	type Config = ();
+	type Info = ();
+	type PostInfo = ();
+	fn dispatch(
+		self,
+		_origin: Self::RuntimeOrigin,
+	) -> crate::DispatchResultWithInfo<Self::PostInfo> {
+		panic!("This implementation should not be used for actual dispatch.");
+	}
+}
+
+/// Runtime Origin which includes a System Origin variant whose `AccountId` is the parameter.
+pub trait AsSystemOriginSigner<AccountId> {
+	/// Extract a reference of the inner value of the System `Origin::Signed` variant, if self has
+	/// that variant.
+	fn as_system_origin_signer(&self) -> Option<&AccountId>;
+}
+
+/// Interface to differentiate between Runtime Origins authorized to include a transaction into the
+/// block, and those who aren't.
+///
+/// Typically, upon validation or application of a transaction, the origin resulting from the
+/// transaction extension (see [`TransactionExtension`]) is checked for authorization. Transaction
+/// is then rejected or applied.
+///
+/// In FRAME, an authorized origin is either an `Origin::Signed` System origin or a custom origin
+/// authorized in a `TransactionExtension`.
+pub trait AsAuthorizedOrigin {
+	/// Whether the origin is authorized to include the transaction in a block.
+	///
+	/// In FRAME returns `false` if the origin is a System `Origin::None` variant, `true` otherwise,
+	/// meaning only signed or custom origin resulting from the transaction extension pipeline are
+	/// authorized.
+	fn is_authorized(&self) -> bool;
+}
```

Comment:
Do we really need this? Maybe just using `Option` or a custom enum to represent the initial `origin` and then at the end just check if this was set?

Reply:
Yes because at the substrate primitives level where `TransactionExtension` is defined, there is no concept of what an origin is, it's just a type. This trait is the way to differentiate between what the implementer decides is the "default" origin for their substrate runtime (if they're using a `frame_system` based runtime, it will probably be `frame_system::RawOrigin::None` BUT not necessarily). `frame_system` and `frame_support` are at a higher level in terms of dependency and we can't use concepts in those crates. As you can see in the `VerifyMultiSignature` implementation now, this trait is useful.

Reply:
Yeah, but as I said the initial origin could just be a `None`? This way, the origin can stay generic and we don't need this trait. Or what do I miss?

Reply:
I see one drawbacks in using `Option<Origin>`, transaction extensions can't add stuff on the origin prior to mutate it to an authorized origin.

Like an extension adding some filter.
Or if `Origin` type get more complex, some additional information on the origin.

That said such extension could always be put after authorizing transaction extensions in the pipeline.
So `Option<Origin>` is also ok to me.

Reply:
I think `Option<Origin>` is redundant because all origin types used in practice, including `frame_system` as well as downstream custom ones, must have a value equivalent to "no origin" already. This would just introduce a "no origin for extension" (`None`) and then possibly a "no origin for call" (`Some(frame_system::RawOrigin::None)`).

In addition to what @gui1117 said, it's just more overhead for people writing extensions because it's explicit in the fn arguments and you have to "unwrap" it all the time and then get your relevant origin. With the trait, the impl is done once and you call it only if you care about this origin being authorized or not. Also, the trait allows users to have more origin values that are not authorized, or values that are authorized only in certain conditions, it's only limited by the trait impl.

Since the origin itself is a generic type, it makes sense we'd have a unifying interface for determining what's authorized and what's not. The trait is simple, the docs are clear, the impl is done for you and you bring it in context only if you use it, I don't see why not use this.

Reply:
When we remove `ValidateUnsigned` we get: bare and general extrinsics start with frame system none origin. Bare extrinsics are only inherents, so they are inherently valid, and nodes include them into a block. 
General extrinsics needs to be authorized to be valid and get into a block.

Seems coherent.

Reply:
`RawOrigin::None` can be a valid origin for inherents for example, after validation.

Reply:
Inherents are not validated by `TransactionExtension`s and for inherents it's not the origin that is "authorized", it's the call itself that is "authorized".

This trait is not about an origin being valid, that's decided by the call logic. An authorized origin means an origin that was authorized by a part of the system through validation and is independent of what the call is, it's a property of the origin itself.

Anyway it feels like this is just a wording issue. I can update the docs to make them more comprehensive or do some renames if necessary.